### PR TITLE
ENH added meg example

### PR DESCRIPTION
**Pull request by QB3**

### CODE REVIEW

**Comment by QB3** [comment]
Somethings interesting in this example is that the initial solution has a sparsity of 1, and the algorithms converge to a solution with a sparsity of 2

**Comment by QB3** [comment]
You can run plot_simulate_evoked_data.py example to see what happends on simulated data @agramfort .

File: examples/meg_lasso_vs_wlassos.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import mne
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:1]: can you not merge this file with the plot_meg one ?

File: examples/meg_lasso_vs_wlassos.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                surf.z[sources], color=(1, 0, 0),
+                scale_factor=s, opacity=1., transparent=True)
+        if save_fname:
+            fname = fig_dir + hemi + save_fname
+            print(fname)
+            if event_id == 1 or event_id == 2:
+                brain.save_montage(fname, order=['lat'])
+            else:
+                brain.save_montage(fname, order=['lat'])
+                # brain.save_montage(fname, order=['ven'])
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:39]: rm

File: examples/meg_lasso_vs_wlassos.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        surf = brain.geo[hemi]
+        sources_h = stc.vertices[i]  # 0 for lh, 1 for rh
+        for sources in sources_h:
+            mlab.points3d(
+                surf.x[sources], surf.y[sources],
+                surf.z[sources], color=(1, 0, 0),
+                scale_factor=s, opacity=1., transparent=True)
+        if save_fname:
+            fname = fig_dir + hemi + save_fname
+            print(fname)
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:34]: rm

File: examples/meg_lasso_vs_wlassos.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            if event_id == 1 or event_id == 2:
+                brain.save_montage(fname, order=['lat'])
+            else:
+                brain.save_montage(fname, order=['lat'])
+                # brain.save_montage(fname, order=['ven'])
+
+            # mlab.savefig(fname)
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:41]: rm

File: examples/meg_lasso_vs_wlassos.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+loose, depth = 0., .8  # corresponds to free orientation
+
+models = ["lasso", "wlasso"]
+for model in models:
+    stc, monitor = apply_solver(
+        evoked, forward, noise_cov, loose, depth, model=model)
+    print("Value of objectives:")
+    print(monitor.objs)
+
+    fig_dir = "../../CD_SUGAR/tex/slides_qbe_long/prebuiltimages/"
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:78]: to change

File: examples/meg_lasso_vs_wlassos.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    save_fname = "%s.png" % model
+
+    plot_blob(stc, fig_dir=fig_dir, save_fname=save_fname)
+
+    ###########################################################################
```
Comment (mathurinm) [inline examples/meg_lasso_vs_wlassos.py:83]: rm or uncomment

File: examples/plot_meg_example.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    M = evoked.data[sel]
+
+    # Whiten data
+    M = np.dot(whitener, M)
+
+    n_orient = 1 if is_fixed_orient(forward) else 3
+
+    X, active_set, monitor = solver(
+        M, gain, n_orient, evoked.nave, p_alpha0=p_alpha0, model=model)
+    # X, active_set, monitor = solver(M, gain, n_orient, )
```
Comment (mathurinm) [inline examples/plot_meg_example.py:88]: rm

File: examples/plot_meg_example.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    X = _reapply_source_weighting(X, source_weighting, active_set)
+
+    # import ipdb; ipdb.set_trace()
```
Comment (mathurinm) [inline examples/plot_meg_example.py:91]: rm

File: examples/plot_meg_example.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+###############################################################################
+# Define your solver
+
+def solver(
+        y_train, X_train, n_orient, nave, p_alpha0=0.7, model="wlasso"):
+    """Run L2 penalized regression and keep 10 strongest locations.
+
+    Parameters
+    ----------
+    M : array, shape (n_channels, n_times)
```
Comment (mathurinm) [inline examples/plot_meg_example.py:107]: y_train and X_train now (putting X before y seems more ML-y to me)

File: examples/plot_meg_example.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        The gain matrix a.k.a. the forward operator. The number of locations
+        is n_dipoles / n_orient. n_orient will be 1 for a fixed orientation
+        constraint or 3 when using a free orientation model.
+    n_orient : int
+        Can be 1 or 3 depending if one works with fixed or free orientations.
+        If n_orient is 3, then ``G[:, 2::3]`` corresponds to the dipoles that
+        are normal to the cortex.
+    nave : int
+        The number of epochs averaged.
+
```
Comment (mathurinm) [inline examples/plot_meg_example.py:119]: monitor_grad is returned

File: examples/plot_meg_example.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    idx_max = np.argmax(np.sum(y_train ** 2, axis=0))
+    y_train = y_train[:, idx_max]
+
+    n_samples, n_features = X_train.shape
+    alpha_max_old = (np.abs(X_train.T @ y_train)).max() / n_samples
+    X_train /= alpha_max_old
+
+    alpha_max = (np.abs(X_train.T @ y_train)).max() / n_samples
+    alpha0 = p_alpha0 * alpha_max
+    # alpha0 = 0.7 * alpha_max
```
Comment (mathurinm) [inline examples/plot_meg_example.py:139]: rm

File: examples/plot_meg_example.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        algo, log_alpha0, monitor_grad, n_outer=n_outer, tol=tol)
+
+    X = criterion.dense0[:, np.newaxis] * np.ones((1, n_times))
+    active_set = criterion.mask0
+    X /= alpha_max_old
+
+    return X, active_set, monitor_grad
+
+
+if __name__ == '__main__':
```
Comment (mathurinm) [inline examples/plot_meg_example.py:166]: remove and unindent

File: examples/plot_simulate_evoked_data.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
+"""
```
Comment (mathurinm) [inline examples/plot_simulate_evoked_data.py:1]: do you need this file ?

**Comment by mathurinm** [comment]
@QB3 do you want to fix the issues and merge, or to close the PR ?

**Comment by agramfort** [comment]
No meg example is sparse-ho ? :(

File: examples/meg_lasso_vs_wlassos.py (hunk 1 above)
Comment (mathurinm) [reply examples/meg_lasso_vs_wlassos.py:1]: @QB3 what do we do with this file, delete or it's needed ? It will not be run by Circle as its name does not start with `plot_*`

**Comment by mathurinm** [comment]
@QB3 I don't know what should be plotted in the `meg_lasso_vs_wlassos.py` example, currently it fails because mayavi is not installed, anyways it's deprecated in favor of pyvistaqt so can you change the script to move to pyvistaqt ?

**Comment by mathurinm** [comment]
ok I give up for today, for some reason the script still won't run; I don't know if it's a matter of backend, I tried to use `pyvistaqt` but it's failing on circle @agramfort

**Comment by mathurinm** [comment]
maybe we should install the mne env with conda rather than a few packages with pip. I went with pip because the mne env takes quite some time to solve.

**Comment by mathurinm** [comment]
@QB3 we have the structure to run on CI, the build only fails because one plot has convergence issues. I'll let you fix locally and finish the PR to have a nice plot

File: doc/doc-requirements.txt [hunk 14]
```diff
@@ -1,4 +1,9 @@
 numpydoc
+mne
```
Comment (mathurinm) [inline doc/doc-requirements.txt:2]: I think we no longer need doc-requirements, they are in the config.yml file now (some with conda, some with pip), WDYT @agramfort ?

**Comment by QB3** [comment]
Good to go for me

**Comment by QB3** [comment]
Good to go for me!

**Comment by agramfort** [comment]
result looks good but I get a lot of warnings running the code. Can you have a look?

```
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
Iteration 1/5 ||Value outer criterion: 3.75e+01 ||norm grad 2.05e+01
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
Iteration 2/5 ||Value outer criterion: 9.09e+00 ||norm grad 4.12e+00
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
Iteration 3/5 ||Value outer criterion: 7.16e+00 ||norm grad 2.72e-01
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
Iteration 4/5 ||Value outer criterion: 5.64e+00 ||norm grad 2.60e-02
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
!!! Inner solver did not converge at epoch 49999, gap: 3.48e-05 > 1.34e-05
!!! Inner solver did not converge at epoch 49999, gap: 4.80e-06 > 3.63e-06
!!! Inner solver did not converge at epoch 49999, gap: 6.24e-07 > 2.99e-07
!!! Inner solver did not converge at epoch 49999, gap: 3.10e-07 > 1.93e-07
!!! Inner solver did not converge at epoch 49999, gap: 3.59e-07 > 9.97e-08
!!! Inner solver did not converge at epoch 49999, gap: 1.68e-07 > 1.10e-07
!!! Inner solver did not converge at epoch 49999, gap: 9.66e-08 > 5.09e-08
!!! Inner solver did not converge at epoch 49999, gap: 5.64e-08 > 2.91e-08
!!! Inner solver did not converge at epoch 49999, gap: 3.25e-08 > 1.69e-08
!!! Inner solver did not converge at epoch 49999, gap: 1.84e-08 > 9.81e-09
!!! Inner solver did not converge at epoch 49999, gap: 1.02e-08 > 5.55e-09
/Users/alex/work/src/scikit-learn/sklearn/linear_model/_base.py:148: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization. If normalization is needed please use sklearn.preprocessing.StandardScaler instead.
  warnings.warn(
!!! Inner solver did not converge at epoch 49999, gap: 5.13e-06 > 5.03e-06
!!! Inner solver did not converge at epoch 49999, gap: 2.78e-06 > 2.41e-06
!!! Inner solver did not converge at epoch 49999, gap: 2.72e-07 > 2.62e-07
!!! Inner solver did not converge at epoch 49999, gap: 1.14e-07 > 8.67e-08
!!! Inner solver did not converge at epoch 49999, gap: 7.16e-08 > 3.50e-08
!!! Inner solver did not converge at epoch 49999, gap: 4.39e-08 > 2.15e-08
!!! Inner solver did not converge at epoch 49999, gap: 2.56e-08 > 1.32e-08
!!! Inner solver did not converge at epoch 49999, gap: 1.44e-08 > 7.69e-09
Iteration 5/5 ||Value outer criterion: 5.06e+00 ||norm grad 4.89e-02
Iteration 1/5 ||Value outer criterion: 3.75e+01 ||norm grad 2.05e+01
Iteration 2/5 ||Value outer criterion: 2.68e+01 ||norm grad 2.20e+01
Iteration 3/5 ||Value outer criterion: 1.59e+01 ||norm grad 7.05e-01
Iteration 4/5 ||Value outer criterion: 1.54e+01 ||norm grad 4.88e-02
Iteration 5/5 ||Value outer criterion: 1.54e+01 ||norm grad 3.42e-03
```

**Comment by QB3** [comment]
`Inner solver did not converge at epoch 49999, gap: 5.13e-06 > 5.03e-06` error messages should be gone.

I currently do not have the other warnings ` FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2. Please leave the normalize parameter to its default value to silence this warning. The default behavior of this estimator is to not do any normalization`.

My version of sklearn is 1.0.1. My version of celer is 0.7dev.

**Comment by mathurinm** [comment]
@agramfort I removed the normalize parameter from celer, maybe you just need to pull celer's master

**Comment by mathurinm** [comment]
meg utils is rendered in the doc. Since there's a single meg example, I believe it can be merged with plot_meg  @QB3 
image

**Comment by mathurinm** [comment]
I don't see the brain in https://984-252772479-gh.circle-artifacts.com/0/dev/auto_examples/plot_meg_lasso_vs_wlasso.html#sphx-glr-auto-examples-plot-meg-lasso-vs-wlasso-py; I believe it would be a better thumbnail than the flat curves we currently have

**Comment by agramfort** [comment]
brains appear as black squares https://988-252772479-gh.circle-artifacts.com/0/dev/auto_examples/plot_meg_lasso_vs_wlasso.html#sphx-glr-auto-examples-plot-meg-lasso-vs-wlasso-py 🤦 

@larsoner do you have any idea why?

**Comment by larsoner** [comment]
My guess is that the latest VTK release is problematic, but haven't looked yet. I'l look this week to see if we have this problem in MNE and fix it there first

File: doc/conf.py [hunk 15]
```diff
@@ -190,11 +190,22 @@
...
 
+scrapers = ('matplotlib',)
+try:
+    import mne
+    import pyvista
+    mne.viz.set_3d_backend('pyvistaqt')
+    pyvista.OFF_SCREEN = False
+    scrapers += ('pyvista',)
+except:
+    pass
```
Comment (larsoner) [inline doc/conf.py:201]: It looks like our docs successfully built Nov 15 and VTK released earlier than that, so it's probably not VTK. I would try using our special Brain scraper:

https://github.com/mne-tools/mne-python/blob/main/doc/conf.py#L353

File: examples/meg_utils.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    - to take into account the noise covariance and do the spatial whitening
+    - to apply loose orientation constraint as MNE solvers
+    - to apply a weighting of the columns of the forward operator as in the
+      weighted Minimum Norm formulation in order to limit the problem
+      of depth bias.
+
+
+    Parameters
+    ----------
+    XXXXXXXXXXXXXXXX Solver is not a parameter of the function, TODO
```
Comment (mathurinm) [inline examples/meg_utils.py:34]: @QB3 can you clean this up ?

**Comment by agramfort** [comment]
success https://989-252772479-gh.circle-artifacts.com/0/dev/auto_examples/index.html !

now clean up time

File: examples/meg_utils.py (hunk 16 above)
Comment (QB3) [reply examples/meg_utils.py]: done

File: doc/conf.py (hunk 15 above)
Comment (QB3) [reply doc/conf.py:201]: Thx a lot @larsoner :)

**Comment by QB3** [comment]
Good to go for me

**Review by agramfort — APPROVED** [review]
Looks great 👍 

@QB3 @mathurinm merge if happy 🙏
